clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -1750,6 +1754,12 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
             llvm::ConstantAsMetadata::get(CGF->Builder.getInt32(
                 ArgInfoAddressSpace(pointeeTy.getAddressSpace()))));
 
+        // Get address qualifier of SYCL kernel pointer parameter from
+        // accessors.
+        argSYCLKernelRuntimeAligned.push_back(
```

Comment:
Quoting the SYCL [spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.parameter.passing) :
`Pointer types are trivially copyable, so they may be passed as kernel parameters. However, only the pointer value itself is passed to the kernel. Dereferencing the pointer on the kernel results in undefined behavior unless the pointer points to an address within a USM memory region that is accessible on the device.`

My understanding from above quote is it is illegal to pass pointers as kernel parameters directly unless it is a USM pointer. So when we are inside this if block  `if (ty->isPointerType())` I assume the pointer is coming from an accessor.

Reply:
Wouldn't `if (ty->isPointerType())` also return true for USM pointers though? This code is in a loop that goes over all the function parameters; When a pointer is seen, I don't see how we can differentiate between a USM pointer and a pointer that's part of the 4 separate kernel arguments that represent the accessor.

Reply:
> When a pointer is seen, I don't see how we can differentiate between a USM pointer and a pointer that's part of the 4 separate kernel arguments that represent the accessor.

Right, I wasn't entirely sure if my assumption was correct here. I will add a USM pointer test case to confirm if the accessor pointers are differentiated from USM pointers.
I had initially considered another approach which was to create and attach a new implicit attribute for pointers from accessors(ex:SYCLAccessorPtrAttr), similar to `SYCLSimdAccessorPtrAttr` or `SYCLAccessorReadonlyAttr` handled in [handleAccessorType](https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L1969)

And do :
```
if(parm->hasAttr<SYCLAccessorPtrAttr>() ) { 
set metadata for kernel-arg-runtime-aligned 
}
```
If there are better approaches to handling this, please let me know.

Reply:
AFAIK `SYCLIntelBufferLocation` attribute is also used for similar purpose.
It is attached in SemaSYCL where we can differentiate accessors from USM pointers and then based on this attribute's value some metadata is attached to the kernel.
Most likely you can't differentiate accessor's pointers from USM pointers on codegen level.
Whether is a good thing to add another attribute or not it is hard for me to say. Probably it is best to ask @AaronBallman .

Reply:
Well this is a bit of a tangled mess, isn't it? :-D I think using an implicit attribute is a viable approach for smuggling this information around the AST. It's a bit of a heavy-handed approach, but it at least means that only SYCL users are paying for the extra memory overhead (as opposed to adding bits onto AST nodes, where everyone pays the cost).